server/paymentProviders/paypal/payment.ts

```diff
@@ -80,33 +81,48 @@ const recordTransaction = async (
   return models.Transaction.createFromContributionPayload(transactionData);
 };
 
-export function recordPaypalSale(order: typeof models.Order, paypalSale): Promise<typeof models.Transaction> {
+// Unfortunately, PayPal has 3 different transaction types: sale, capture and transaction. Though they all
+// represent the same thing (IDs point toward the same thing), the data structure is different. We thus need
+// these 3 functions to handle them all.
+
+export function recordPaypalSale(
+  order: typeof models.Order,
+  paypalSale: PaypalSale,
+): Promise<typeof models.Transaction> {
   const currency = paypalSale.amount.currency;
   const amount = paypalAmountToCents(paypalSale.amount.total);
   const fee = paypalAmountToCents(get(paypalSale, 'transaction_fee.value', '0.0'));
-  return recordTransaction(order, amount, currency, fee, { data: { paypalSale } });
+  return recordTransaction(order, amount, currency, fee, {
+    data: { paypalSale, paypalCaptureId: paypalSale.id },
+  });
 }
 
 export function recordPaypalTransaction(
   order: typeof models.Order,
-  paypalTransaction,
+  paypalTransaction: PaypalTransaction,
   { data = undefined, createdAt = undefined } = {},
 ): Promise<typeof models.Transaction> {
   const currency = paypalTransaction.amount_with_breakdown.gross_amount.currency_code;
   const amount = floatAmountToCents(parseFloat(paypalTransaction.amount_with_breakdown.gross_amount.value));
   const fee = parseFloat(get(paypalTransaction.amount_with_breakdown, 'fee_amount.value', '0.0'));
-  return recordTransaction(order, amount, currency, fee, { data: { ...data, paypalTransaction }, createdAt });
+  return recordTransaction(order, amount, currency, fee, {
+    data: { ...data, paypalTransaction, paypalCaptureId: paypalTransaction.id },
+    createdAt,
+  });
 }
 
 export const recordPaypalCapture = async (
   order: typeof models.Order,
-  capture,
+  capture: PaypalCapture,
   { data = undefined, createdAt = undefined } = {},
 ): Promise<typeof models.Transaction> => {
   const currency = capture.amount.currency_code;
   const amount = paypalAmountToCents(capture.amount.value);
   const fee = paypalAmountToCents(get(capture, 'seller_receivable_breakdown.paypal_fee.value', '0.0'));
-  return recordTransaction(order, amount, currency, fee, { data: { ...data, capture }, createdAt });
+  return recordTransaction(order, amount, currency, fee, {
+    data: { ...data, capture, paypalCaptureId: capture.id },
```

Comment:
Are we sure these IDs are not going to collide? Should we add a prefix based on the type of transaction?

Reply:
They're not supposed to since they all represent the same thing. In any case, it's not a unique index, so it won't crash if that ever happens.